File: java/dsu.java
class dsu{
    private int[] parent;
    private int[] sz;

    public dsu(int n){
        parent = new int[n];
        sz = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            sz[i] = 1;
        }
    }

    public int root(int x) {
        if (parent[x] == x) return x;
        return parent[x] = root(parent[x]);
    }

    public void merge(int x, int y){
        x = root(x);
        y = root(y);
        if (x == y) return;
        if (sz[x] < sz[y]) {
            int tmp = x;
            x = y;
            y = tmp;
        }
        parent[y] = x;
        sz[x] += sz[y];
    }

    public boolean same(int x, int y){
        return root(x) == root(y);
    }

    public int size(int x){
        return sz[root(x)];
    }
}

File: py/dsu.py
class dsu:
    def __init__(self, n):
        self.parent = list(range(n))
        self.sz = [1]*n

    def root(self, x):
        if self.parent[x] == x:
            return x
        self.parent[x] = self.root(self.parent[x])
        return self.parent[x]

    def merge(self, x, y):
        x = self.root(x)
        y = self.root(y)
        if x == y:
            return
        if self.sz[x] < self.sz[y]:
            x, y = y, x
        self.parent[y] = x
        self.sz[x] += self.sz[y]

    def same(self, x, y):
        return self.root(x) == self.root(y)

    def size(self, x):
        return self.sz[self.root(x)]